src/model/liveselection.js

```diff
@@ -0,0 +1,445 @@
+ * {@link engine.model.Document document} and has it ranges updated accordingly. Internal implementation of this
+ * mechanism bases on {@link engine.model.LiveRange live ranges}.
+ *
+ * Differences between {@link engine.model.Selection} and `LiveSelection` are three:
+ * * there is always a range in `LiveSelection`, even if no ranges were added - in this case, there is a
+ * "default range" in selection which is a collapsed range set at the beginning of the {@link engine.model.Document document},
+ * * ranges added to this selection updates automatically when the document changes,
+ * * live selection may have attributes.
+ *
+ * @memberOf engine.model
+ */
+export default class LiveSelection extends Selection {
+	/**
+	 * Creates an empty document selection for given {@link engine.model.Document}.
+	 *
+	 * @param {engine.model.Document} document Document which owns this selection.
+	 */
+	constructor( document ) {
+		super();
+
+		/**
+		 * Document which owns this selection.
+		 *
+		 * @private
+		 * @member {engine.model.Document} engine.model.Selection#_document
+		 */
+		this._document = document;
+
+		/**
+		 * List of attributes set on current selection.
+		 *
+		 * @protected
+		 * @member {Map} engine.model.LiveSelection#_attrs
+		 */
+		this._attrs = new Map();
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	get isCollapsed() {
+		const length = this._ranges.length;
+
+		return length === 0 ? true : super.isCollapsed;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	get anchor() {
+		return super.anchor || this._getDefaultRange().start;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	get focus() {
+		return super.focus || this._getDefaultRange().start;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	get rangeCount() {
+		return this._ranges.length ? this._ranges.length : 1;
+	}
+
+	/**
+	 * Unbinds all events previously bound by document selection.
+	 */
+	destroy() {
+		for ( let i = 0; i < this._ranges.length; i++ ) {
+			this._ranges[ i ].detach();
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	*getRanges() {
+		if ( this._ranges.length ) {
+			yield *super.getRanges();
+		} else {
+			yield this._getDefaultRange();
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	getFirstRange() {
+		return super.getFirstRange() || this._getDefaultRange();
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	removeAllRanges() {
+		this.destroy();
+		super.removeAllRanges();
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	setRanges( newRanges, isLastBackward ) {
+		this.destroy();
+		super.setRanges( newRanges, isLastBackward );
+	}
+
+	/**
+	 * Creates an instance of {@link engine.model.Selection} that has same ranges and direction as this selection. Since
+	 * this will be instance of `Selection` instead of `LiveSelection`, it will not be automatically updated or rendered,
+	 * so it can be used in algorithms using and modifying selection.
+	 *
+	 * @returns {engine.model.Selection} Selection instance which ranges and direction is equal to this selection.
+	 */
+	getSnapshot() {
```

Comment:
We have `Position.createFromPosition`, `Range.createFromRange` so why not `Selection.createFromSelection`?